contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -1231,3 +1259,343 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+		stmts = parsetree_nth_stmt(stmt_list, i++);
+		update_AlterDefaultPrivilegesStmt(stmts, schema_name, schema_owner, dbo_user, db_datawriter, privilege_to_string(ACL_DELETE));
+
+		/* Run all subcommands */
+		foreach(parsetree_item, stmt_list)
+		{
+			Node		*stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+			PlannedStmt *wrapper;
+
+			/* need to make a wrapper PlannedStmt */
+			wrapper = makeNode(PlannedStmt);
+			wrapper->commandType = CMD_UTILITY;
+			wrapper->canSetTag = false;
+			wrapper->utilityStmt = stmt;
+			wrapper->stmt_location = 0;
+			wrapper->stmt_len = 0;
+
+			/* do this step */
+			ProcessUtility(wrapper,
+						"(CREATE DATABASE ROLES) ",
+						false,
+						PROCESS_UTILITY_SUBCOMMAND,
+						NULL,
+						NULL,
+						None_Receiver,
+						NULL);
+		}
+		tuple = heap_getnext(tblscan, ForwardScanDirection);
+	}
+
+	/* Cleanup. */
+	table_endscan(tblscan);
+	table_close(namespace_rel, RowExclusiveLock);
+}
+
+static void
+create_db_roles_if_not_exists(const uint16 dbid,
+							const char *dbname)
+{
+	StringInfoData  query;
+	Oid     datdba;
+	const char		*prev_current_user;
+	uint16			old_dbid;
+	const char		*old_dbname;
+	const char		*db_datareader;
+	const char 		*db_datawriter;
+	ListCell		*parsetree_item;
+	List			*stmt_list;
+	Node			*stmts;
+	int			i=0;
+
+	/*
+	 * During upgrade, the migration mode is reset to single-db so we cannot
+	 * call get_physical_user_name() directly. Detect whether the original
+	 * migration was single-db or multi-db.
+	 */
+	MigrationMode baseline_mode = is_user_database_singledb(dbname) ? SINGLE_DB : MULTI_DB;
+
+	db_datareader = get_db_datareader_name_by_mode(dbname, baseline_mode);
+	db_datawriter = get_db_datawriter_name_by_mode(dbname, baseline_mode);
+
+	/*
+	 * database roles prepends dbname based on single-db or multi-db. If for
+	 * some reason database roles exist, then that is a bigger problem.
+	 * The roles should be manually cleaned up before the upgrade.
+	 */
+	if (OidIsValid(get_role_oid(db_datareader, true)))
+	{
+		ereport(LOG,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("role \"%s\" already exists. Please drop the role and restart upgrade.", db_datareader)));
+		return;
+	}
+
+	if (OidIsValid(get_role_oid(db_datawriter, true)))
+	{
+		ereport(LOG,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("role \"%s\" already exists. Please drop the role and restart upgrade.", db_datawriter)));
+		return;
+	}
+
+	datdba = get_role_oid("sysadmin", false);
+	check_can_set_role(GetSessionUserId(), datdba);
+
+	initStringInfo(&query);
+	appendStringInfo(&query, "CREATE ROLE dummy INHERIT; ");
+	appendStringInfo(&query, "CREATE ROLE dummy INHERIT; ");
+
+	stmt_list = raw_parser(query.data, RAW_PARSE_DEFAULT);
+	Assert(list_length(stmt_list) == 2);
+
+	/* Set current user to session user for create permissions */
+	prev_current_user = GetUserNameFromId(GetUserId(), false);
+	bbf_set_current_user("sysadmin");
```

Comment:
We would still want `bbf_role_admin` to be the admin of these fixed roles right? Are we making sure of it?